pkg/builder/controller.go

```diff
@@ -52,21 +52,32 @@ const (
 )
 
 // Builder builds a Controller.
-type Builder struct {
+type Builder = TypedBuilder[reconcile.Request]
+
+// TypedBuilder builds a Controller. The request is the request type
+// that is passed to the workqueue and then to the Reconciler.
+// The workqueue de-duplicates identical requests.
+type TypedBuilder[request comparable] struct {
 	forInput         ForInput
 	ownsInput        []OwnsInput
-	rawSources       []source.Source
-	watchesInput     []WatchesInput
+	rawSources       []source.TypedSource[request]
+	watchesInput     []WatchesInput[request]
 	mgr              manager.Manager
 	globalPredicates []predicate.Predicate
```

Comment:
The predicate type is the type of the object. In order to be able to watch multiple sources with different types, the concrete type of any source must not leak into the builder and hence everything in the builder related to the object remains typed as client.Object. That was basically the motivation for the interface changes in https://github.com/kubernetes-sigs/controller-runtime/pull/2783

Reply:
You probably ran into this when looking at `TypedPredicates`. The only reason they have a type is to be able to implement `ApplyToWatches(opts *WatchesInput[request])` because when referencing another generic type, you have to specify the generic, you can not say "applies regardless of generic type". The predicate itself doesn't care at all and the predicate it sets on `WatchesInput` is typed to `client.Object`. I'll have to think about this a bit if I can come up with something better, but the only alternative I see is to change the `ApplyToWatches` interface to apply to an interface implemented by `WatchesInput` rather than the concrete `WatchesInput` as that is the only way to reference it regardless of generic.

Reply:
I got rid of the generics in `Predicate` and `Projection` in 10a38deccfbdd15d14916ae89222a023b5d09b40, let em know what you think. Will squash at the end.

Reply:
Looks good to me!